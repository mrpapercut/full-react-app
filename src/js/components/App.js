import {Component, createElement as E} from 'react';
import {connect} from 'react-redux';
import {Route, Switch, withRouter} from 'react-router-dom';
import loadable from '@loadable/component';

import {
    isLoading
} from '~/actions/mainActions';

import createElement from '~/util/createElement';

const Sidebar           = loadable(() => import(/* webpackChunkName: "Sidebar" */'~/components/Sidebar'));
const MainComponent     = loadable(() => import(/* webpackChunkName: "MainComponent" */'~/components/MainComponent'));
const ImageComponent    = loadable(() => import(/* webpackChunkName: "ImageComponent" */'~/components/ImageComponent'));
const StateComponent    = loadable(() => import(/* webpackChunkName: "StateComponent" */'~/components/StateComponent'));
const StoreComponent    = loadable(() => import(/* webpackChunkName: "StoreComponent" */'~/components/StoreComponent'));
const NotFound404       = loadable(() => import(/* webpackChunkName: "NotFound404" */'~/components/NotFound404'));

const [div] = ['div'].map(createElement);

class App extends Component {
    constructor(props) {
        super(props);

        this.state = this.getInitialState();
    }

    getInitialState() {
        return {
            isLoading: this.props.isLoading
        };
    }

    render() {
        const unmappedRoutes = [{
            path: '/image',
            component: ImageComponent
        }, {
            path: '/state',
            component: StateComponent
        }, {
            path: '/store',
            component: StoreComponent
        }, {
            path: '*',
            component: NotFound404
        }];

        const mappedRoutes = unmappedRoutes.map((r, key) => {
            return E(Route, {
                path: r.path,
                render: () => E(r.component),
                key
            });
        });

        return [
            this.state.isLoading ? div({key: 'loader', id: 'loader'}) : null,
            E(Sidebar, {key: 'sidebar'}),
            div({
                key: 'mainwindow',
                className: 'mainwindow'
            },
                E(Switch, {
                    key: 'mainSwitch'
                },
                    E(Route, {
                        key: 'mainComponent',
                        path: '/',
                        exact: true,
                        render: () => E(MainComponent)
                    }),
                    mappedRoutes
                )
            )
        ];
    }
}

const mapStateToProps = ({mainReducer}) => {
    return {
        isLoading: mainReducer.isLoading
    };
};

const mapDispatchToProps = dispatch => {
    return {
        setLoading: bool => dispatch(isLoading(bool))
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));
