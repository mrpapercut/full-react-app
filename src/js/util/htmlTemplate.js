const htmlTemplate = ({title, body, scriptTags, preloadedState}) => `<!doctype html>
<html>
<head>
    <meta charset="utf8"></meta>
    <title>${title}</title>
    <base href="/">

    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>
    <div id="appwrapper">${body}</div>
    <script>${preloadedState}</script>
    ${scriptTags}
</body>
</html>
`;

export default htmlTemplate;
