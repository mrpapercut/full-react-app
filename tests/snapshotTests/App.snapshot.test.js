import {createElement} from 'react';
import renderer from 'react-test-renderer';

import {wrapElementWithRouter} from '../util/snapshot-test-util';

import App from '~/components/App';

test('App', () => {
    const component = renderer.create(wrapElementWithRouter(createElement(App)));
    expect(component.toJSON()).toMatchSnapshot();
});

test('App with loader enabled', () => {
    const component = renderer.create(wrapElementWithRouter(createElement(App), {
        mainReducer: {
            isLoading: true
        }
    }));
    expect(component.toJSON()).toMatchSnapshot();
});

test('App with loader disabled', () => {
    const component = renderer.create(wrapElementWithRouter(createElement(App), {
        mainReducer: {
            isLoading: false
        }
    }));
    expect(component.toJSON()).toMatchSnapshot();
});
