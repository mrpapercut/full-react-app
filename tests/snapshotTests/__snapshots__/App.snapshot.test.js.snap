// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 1`] = `
<div
  className="mainwindow"
/>
`;

exports[`App with loader disabled 1`] = `
<div
  className="mainwindow"
/>
`;

exports[`App with loader enabled 1`] = `
Array [
  <div
    id="loader"
  />,
  <div
    className="mainwindow"
  />,
]
`;
